recipe/patches/0001-Support-both-clang-17-and-clang-18-ABI.patch

```diff
@@ -0,0 +1,81 @@
+From 3fb1af546ca2e74ea6dfe5478e78c4981481740a Mon Sep 17 00:00:00 2001
+From: Isuru Fernando <isuruf@gmail.com>
+Date: Fri, 3 Jan 2025 22:06:55 +0530
+Subject: [PATCH] Support both clang-17 and clang-18 ABI
+
+---
+ CMake/AbseilDll.cmake             |  1 +
+ absl/log/CMakeLists.txt           |  3 +++
+ absl/log/internal/log_message2.cc | 27 +++++++++++++++++++++++++++
+ 3 files changed, 31 insertions(+)
+ create mode 100644 absl/log/internal/log_message2.cc
+
+diff --git a/CMake/AbseilDll.cmake b/CMake/AbseilDll.cmake
+index a3fb15a1..bbc41128 100644
+--- a/CMake/AbseilDll.cmake
++++ b/CMake/AbseilDll.cmake
+@@ -183,6 +183,7 @@ set(ABSL_INTERNAL_DLL_FILES
+   "log/internal/log_format.h"
+   "log/internal/log_impl.h"
+   "log/internal/log_message.cc"
++  "log/internal/log_message2.cc"
+   "log/internal/log_message.h"
+   "log/internal/log_sink_set.cc"
+   "log/internal/log_sink_set.h"
+diff --git a/absl/log/CMakeLists.txt b/absl/log/CMakeLists.txt
+index 5b9a5f9d..f23d9105 100644
+--- a/absl/log/CMakeLists.txt
++++ b/absl/log/CMakeLists.txt
+@@ -187,6 +187,7 @@ absl_cc_library(
+     log_internal_message
+   SRCS
+     "internal/log_message.cc"
++    "internal/log_message2.cc"
+   HDRS
+     "internal/log_message.h"
+   COPTS
+@@ -220,6 +221,8 @@ absl_cc_library(
+     absl::time
+ )
+ 
++set_source_files_properties(internal/log_message2.cc PROPERTIES COMPILE_FLAGS -fclang-abi-compat=17)
```

Comment:
Could we patch in the clang 17 symbols without changing the default mangling, rather than using the clang 17 mangling and add the clang 18 style symbols? That would make it easier to transition in the future (or we have to remember to remove this bit for the next version & migration)

Reply:
I don't understand. Can you explain?

Reply:
Ping on this.

Reply:
Ok, let's stop the bleeding here first.

To answer your question: there's two manglings, the clang 17 and clang 18 one. Only one gets generated from the sources, and for compatibility with both we need to add an implementation of the other.

AFAIU we can either let clang generate the clang 17 mangling and manually add the clang 18 signature (as you do here), or we could could let clang 18 produce its default mangling and manually add the clang 17 signature

Reply:
We let clang 18 produce its default mangling with `log_message.cc` and we manually add clang 17 signature with `log_message2.cc`

Reply:
Ah, I hadn't noticed on the phone that you're only setting the compat ABI only for one file. I had been thinking about manually providing the symbols with the right signature, but your approach is better. 👍